executor/common_linux.h

```diff
@@ -4103,3 +4103,281 @@ static void setup_usb()
 // The macro is used in generated C code.
 #define CAST(f) ({void* p = (void*)f; p; })
 #endif
+
+#if SYZ_EXECUTOR || __NR_syz_mount_fuse
+#include <fcntl.h>
+#include <linux/fuse.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+/* Perform correct FUSE initialization:
+ * 1) Create target directory if missing
+ * 2) Mount FUSE fs on top of target directory
+ * 3) Return fd referring to the target directory
+ */
+static volatile long syz_mount_fuse(volatile long a0, /* /dev/fuse fd. */
+				    volatile long a1, /* Mount target. */
+				    volatile long a2) /* Mount flags. */
+{
+	const char* file_system_type = "fuse";
+	const char* mount_source = "fuse"; /* Not used by FUSE. */
+	char* mount_target;
+	char opts[128];
+	unsigned long mount_flags;
+	struct stat stbuf;
+	int fd_fuse;
+	int ret;
+
+	fd_fuse = (int)a0;
+	mount_target = (char*)a1;
+	mount_flags = (unsigned long)a2;
+
+	ret = mkdir(mount_target, S_IRWXU);
```

Comment:
I don't see what's here cannot be done in descriptions. Am I missing something? Have you tried to express all of this in descriptions? 
See sys/linux/filesystem.txt for now to form mount options.

Reply:
The important bit about this syscalls is the returned file descriptor. It refers to the directory where the FUSE filesystem has been mounted, and allows the syscalls that follow to target files stored on a FUSE filesystem without hardcoding paths (using `*at()` syscalls and passing this fd). From what I have seen, there isn't a real syscall that supports this behavior.

Regarding fuzzing the mount options, a discriminated syscall for FUSE is already present inside [sys/linux/fs_fuse.txt](https://github.com/google/syzkaller/blob/master/sys/linux/fs_fuse.txt#L16).

Reply:
We don't need a pseudo-syscall for this. An openat specialization can accept and return a more specialized fd type.

Reply:
Oh, wait, I was thinking about something else (the other pseudo-syscall).

Reply:
I still don't see what we gain much here. All scenarios that are possible with these pseudo-syscalls seem to be easily possible with mkdir/openat as well. It maybe will shift the focus just a bit, but it does not seem to justify more custom C code, more maintainable, larger reproducers. "Slightly shifting focus" may be useful for all of 10000 kernel subsystems, but we can't add 10000 pseudo-syscalls just for that.
If you really want to improve this, we need to think how syzlang needs to be changed to support such richer resource annotations.

Reply:
I have tried and wasn't able to reproduce FUSE file interaction without this and syz_handle_fuse_req() (with this being strictly needed for interacting with directories).

My guess it that without returning the fd from the mount() syscall, syzkaller is free to execute the mount() and mkdir()/openat() on 2 different threads since there isn't any dependency between the two syscalls. This leads to a data race, and can end up with mkdir()/openat() executing before mount() and targeting the previously mounted filesystem (e.g. ext4). Please, correct me if what I'm saying is wrong.

Reply:
Coming back to this after a few days of fuzzing. Using only mkdir/openat$dir/mount$fuse, the fuzzer hasn't been able to cover all the code it was covering with the pseudo-syscall syz_fuse_mount(). The main problem seems to be opening "FUSE directories" because `fs/fuse/readdir.c` hasn't been touched at all (previously ~72% coverage).

If you are still against this pseudo-syscall, I can generalize it to something like syz_mount(), accepting the same parameters as a real mount(), but returning the mounted directory fd. This way, all filesystems can use it, and for FUSE I can hardcode the parameters needed in a syscall specialization. Would that be acceptable?

Reply:
Interesting. I wonder what's the reason for no coverage in fs/fuse/readdir.c.
I am now trying to re-create the experiment.

Reply:
I also need to enable openat$fuse

Reply:
also these:
```
		"read$FUSE",
		"write$FUSE*",
		"syz_fuse_handle_req",
		"getuid",
```

Reply:
Did you enable getdents? That's what I think will give coverage in fs/fuse/readdir.c.

Reply:
> Coming back to this after a few days of fuzzing. Using only mkdir/openat$dir/mount$fuse, the fuzzer hasn't been able to cover all the code it was covering with the pseudo-syscall syz_fuse_mount().

Sorry, I didn't express myself too well. With "Using only mkdir/openat$dir/mount$fuse" I meant these instead of syz_fuse_mount (combined by the fuzzer they should substitute syz_fuse_mount). You still need to enable VFS/FUSE related syscalls to reproduce some interaction.

Reply:
> Did you enable getdents? That's what I think will give coverage in fs/fuse/readdir.c.

Yes. Just to be sure, I am rerunning syzkaller with the following config:
```
"disable_syscalls": [
                "fsetxattr$security_capability",
                "fsetxattr$security_evm",
                "fsetxattr$security_ima",
                "fsetxattr$security_selinux",
                "fsetxattr$system_posix_acl",
                "fsetxattr$trusted_overlay_nlink",
                "fsetxattr$trusted_overlay_opaque",
                "fsetxattr$trusted_overlay_origin",
                "fsetxattr$trusted_overlay_redirect",
                "fsetxattr$trusted_overlay_upper",
                "mkdirat$cgroup",
                "mkdirat$cgroup_root",
                "openat$autofs",
                "openat$bsg",
                "openat$cgroup",
                "openat$cgroup_devices",
                "openat$cgroup_freezer_state",
                "openat$cgroup_int",
                "openat$cgroup_netprio_ifpriomap",
                "openat$cgroup_procs",
                "openat$cgroup_ro",
                "openat$cgroup_root",
                "openat$cgroup_subtree",
                "openat$cgroup_type",
                "openat$full",
                "openat$hpet",
                "openat$hwrng",
                "openat$loop_ctrl",
                "openat$md",
                "openat$null",
                "openat$nvram",
                "openat$pidfd",
                "openat$procfs",
                "openat$ptmx",
                "openat$ptp0",
                "openat$random",
                "openat$rfkill",
                "openat$rtc",
                "openat$selinux_attr",
                "openat$snapshot",
                "openat$sndseq",
                "openat$sndtimer",
                "openat$thread_pidfd",
                "openat$ttyS3",
                "openat$urandom",
                "openat$vcs",
                "openat$vcsa",
                "openat$vcsu",
                "openat$vga_arbiter",
                "openat$zero",
                "read$ptp",
                "read$rfkill",
                "read$snapshot",
                "write$binfmt_aout",
                "write$binfmt_elf32",
                "write$binfmt_elf64",
                "write$binfmt_misc",
                "write$binfmt_script",
                "write$cgroup_devices",
                "write$cgroup_freezer_state",
                "write$cgroup_int",
                "write$cgroup_netprio_ifpriomap",
                "write$cgroup_pid",
                "write$cgroup_subtree",
                "write$cgroup_type",
                "write$rfkill",
                "write$selinux_attr",
                "write$snapshot",
                "write$sndseq"
        ],
        "enable_syscalls": [
                "openat$fuse",
                "openat$cuse",
                "openat$dir",
                "openat",
                "close",
                "mount$fuse",
                "mount$fuseblk",
                "umount2",
                "ioctl$FUSE_DEV_IOC_CLONE",
                "ioctl$FIBMAP",
                "write",
                "read",
                "syz_fuse_handle_req",
                "mkdirat",
                "mkdir",
                "renameat",
                "ftruncate",
                "fallocate",
                "syncfs",
                "fdatasync",
                "getdents",
                "getdents64",
                "readlinkat",
                "unlinkat",
                "symlinkat",
                "linkat",
                "sync_file_range",
                "splice",
                "poll",
                "fsync",
                "fchmod",
                "fstatfs",
                "fstat",
                "fgetxattr",
                "fsetxattr",
                "flistxattr",
                "fremovexattr",
                "mmap",
                "lseek",
                "flock",
                "fcntl$lock",
                "getgid",
                "getuid",
                "getpid"
        ]
```

Reply:
> Just to be sure, I am rerunning syzkaller with the following config: [...]

Reconfirming that after ~260 hours of fuzzing `fs/fuse/readdir.c` has no coverage.

Reply:
Weird. I hate writing this off to "fuzzer not being able to open the dir", because it's really just 1 additional openat call. There may be some more fundamental issue here. But I got distracted from chasing it further yesterday...

Reply:
> Weird. I hate writing this off to "fuzzer not being able to open the dir", because it's really just 1 additional openat call.

I agree with that, but currently I don't know why it is not happening. To make sure getdents() can be called, I have written this syzkaller program (which does not use syz_fuse_mount()), and it works:
```
mkdir(&(0x7f0000000100)='./file0\x00', 0x0)
r0 = openat$fuse(0xffffffffffffff9c, &(0x7f0000000000)='/dev/fuse\x00', 0x2, 0x0)
mount$fuse(0x0, &(0x7f0000000040)='./file0\x00', &(0x7f0000000080)='fuse\x00', 0x0, &(0x7f0000000bc0)={{'fd', 0x3d, r0}, 0x2c, {'rootmode', 0x3d, 0x4000}, 0x2c, {'user_id'}, 0x2c, {'group_id'}})
read$FUSE(r0, &(0x7f0000000c80)={0x2020, 0x0, <r1=>0x0}, 0x2020)
write$FUSE_INIT(r0, &(0x7f0000000140)={0x50, 0x0, r1}, 0x50)
r3 = openat$dir(0xffffffffffffff9c, &AUTO='./file0\x00', 0x0, 0x0)
syz_fuse_handle_req(r0, &(0x7f00000061c0)=""/8192, 0x2000, &(0x7f0000000c00)={0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, &(0x7f0000000100)={0x20, 0x0, 0x0, {0x0, 0x8}}, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0})
getdents(r3, &AUTO=""/1024, AUTO)
syz_fuse_handle_req(r0, &(0x7f0000006180)=""/8192, 0x2000, &(0x7f0000000880)={0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, &(0x7f00000005c0)={0x30, 0x0, 0x0, [{0x0, 0x0, 0x4, 0x0, '#,,-'}]}, 0x0, 0x0, 0x0})
```

The syscalls I use are all enabled in the above configs, so the fuzzer should be able to come up with this.

Taking a quick glance at the programs generated by the fuzzer, I have noticed that when it calls opeat$dir() it only tries to open nested directories and not the FUSE root dir (and this requires a lot more exchanges between FUSE kernel and FUSE daemon). Maybe that is related to the issue.

Reply:
Running with these enabled syscalls:
```
	"enable_syscalls": [
		"mount$fuse",
		"mkdirat",
		"openat",
		"write",
		"read",
		"getdents",
		"openat$fuse",
		"openat$dir",
		"read$FUSE",
		"write$FUSE*",
		"syz_fuse_handle_req",
		"getuid"
	],
```
for:
```
uptime | 20h2m10s
fuzzing | 588h30m0s
```
I've got 41% coverage in fs/fuse/readdir.c

Reply:
I've tried to enforce O_CREAT for all openat's with:
```
--- a/sys/linux/init.go
+++ b/sys/linux/init.go
@@ -14,6 +14,7 @@ func InitTarget(target *prog.Target) {
        arch := &arch{
                unix:                        targets.MakeUnixNeutralizer(target),
                clockGettimeSyscall:         target.SyscallMap["clock_gettime"],
+               O_CREAT:                     target.GetConst("O_CREAT"),
                MREMAP_MAYMOVE:              target.GetConst("MREMAP_MAYMOVE"),
                MREMAP_FIXED:                target.GetConst("MREMAP_FIXED"),
                SYSLOG_ACTION_CONSOLE_OFF:   target.GetConst("SYSLOG_ACTION_CONSOLE_OFF"),
@@ -113,6 +114,7 @@ type arch struct {
 
        clockGettimeSyscall *prog.Syscall
 
+       O_CREAT                     uint64
        MREMAP_MAYMOVE              uint64
        MREMAP_FIXED                uint64
        SYSLOG_ACTION_CONSOLE_OFF   uint64
@@ -149,6 +151,8 @@ type arch struct {
 func (arch *arch) neutralize(c *prog.Call) {
        arch.unix.Neutralize(c)
        switch c.Meta.CallName {
+       case "openat":
+               c.Args[2].(*prog.ConstArg).Val |= arch.O_CREAT
        case "mremap":
                // Add MREMAP_FIXED flag, otherwise it produces non-deterministic results.
                flags := c.Args[3].(*prog.ConstArg)
```

but after I restarted the same instances, all coverage in readdir.c disappeared (though, majority of coverage in fuse stayed). Not sure if it's something fundamental, or it just needs to create new reproducers that will work with O_CREAT.

Reply:
After reverting the O_CREAT patch, I get 46% readdir.c coverage within 1.5h.

Reply:
> Running with these enabled syscalls:
> 
> ```
> 	"enable_syscalls": [
> 		"mount$fuse",
> 		"mkdirat",
> 		"openat",
> 		"write",
> 		"read",
> 		"getdents",
> 		"openat$fuse",
> 		"openat$dir",
> 		"read$FUSE",
> 		"write$FUSE*",
> 		"syz_fuse_handle_req",
> 		"getuid"
> 	],
> ```
> 
> for:
> 
> ```
> uptime | 20h2m10s
> fuzzing | 588h30m0s
> ```
> 
> I've got 41% coverage in fs/fuse/readdir.c

I have checked my instance over the weekend as well, and confirm that I had ~43% coverage in `fs/fuse/readdir.c` as well after ~600 hours of fuzzing.

> After reverting the O_CREAT patch, I get 46% readdir.c coverage within 1.5h.

Is the fuzzer reusing the previously dropped programs (due to the O_CREAT patch introduction)?

Reply:
> > After reverting the O_CREAT patch, I get 46% readdir.c coverage within 1.5h.
> 
> Is the fuzzer reusing the previously dropped programs (due to the O_CREAT patch introduction)?

Hard to say.
After each restart it triages all programs in the corpus and throws away the ones that don't give new coverage. So it could throw away lots of them. However, it could keep them if they give new coverage in other parts of the kernel. So maybe it preserved just enough to regain majority of coverage in readdir.